components/focus-trap/focus-trap-mixin.js

```diff
@@ -0,0 +1,99 @@
+import { getNextFocusable, getPreviousFocusable } from '../../helpers/focus.js';
+import { html } from 'lit-element/lit-element.js';
+import { isComposedAncestor } from '../../helpers/dom.js';
+
+export const FocusTrapMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			trap: { type: Boolean, reflect: true }
+		};
+	}
+
+	constructor() {
+		super();
+		this.trap = false;
+		this._handleFocusTrapBodyFocus = this._handleFocusTrapBodyFocus.bind(this);
+	}
+
+	async attributeChangedCallback(name, oldval, newval) {
+		super.attributeChangedCallback(name, oldval, newval);
+		if (name !== 'trap') return;
+		if (this.trap) {
+			document.body.addEventListener('focus', this._handleFocusTrapBodyFocus, true);
+		} else {
+			document.body.removeEventListener('focus', this._handleFocusTrapBodyFocus, true);
+		}
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		if (this.trap) {
+			document.body.removeEventListener('focus', this._handleFocusTrapBodyFocus, true);
+		}
+	}
+
+	_focusTrapFocusFirst() {
+		if (this._focusFirst) {
+			this._focusFirst();
+			return;
+		}
+		const focusable = getNextFocusable(this.shadowRoot.querySelector('.d2l-focus-trap-start'));
+		if (focusable) focusable.focus();
+	}
+
+	_getFocusTrapContainer() {
+		return this.shadowRoot.querySelector('.d2l-focus-trap-start').parentNode;
+	}
+
+	_handleFocusTrapBodyFocus(e) {
+		const trapContainer = this._getFocusTrapContainer();
+		const target = e.composedPath()[0];
+		if (isComposedAncestor(trapContainer, target)) return;
+		this._focusTrapFocusFirst();
+	}
+
+	_handleFocusTrapEndFocusIn(e) {
+		const trapContainer = this._getFocusTrapContainer();
+		if (isComposedAncestor(trapContainer, e.relatedTarget)) {
+			// user is exiting trap via forward tabbing...
+			const firstFocusable = getNextFocusable(this.shadowRoot.querySelector('.d2l-focus-trap-start'));
+			if (firstFocusable) {
+				firstFocusable.focus();
+				return;
+			}
+		}
+		this._focusTrapFocusFirst();
+	}
+
+	_handleFocusTrapStartFocusIn(e) {
+		const trapContainer = this._getFocusTrapContainer();
+		if (isComposedAncestor(trapContainer, e.relatedTarget)) {
+			// user is exiting trap via back tabbing...
+			const lastFocusable = getPreviousFocusable(this.shadowRoot.querySelector('.d2l-focus-trap-end'));
+			if (lastFocusable) {
+				lastFocusable.focus();
+				return;
+			}
+		}
+		this._focusTrapFocusFirst();
+	}
+
+	_renderFocusTrap(inner) {
+		if (this.trap) {
+			return html`
```

Comment:
The only difference here is the absence of `tabindex="0"`.  I'd like to find a better way to avoid this duplication.  I thought of setting `tabindex="-1"`, but that would still make it focusable via script and may be targeted by various getFocusable helpers. Omitting it is really what I want when not actively trapping.

Reply:
Can you use the `ifDefined` thing? Like:

```
const tabindex = this.trap ? "0" : undefined;
const blah = html`<span tabindex="${ifDefined(tabindex)}">`;

Reply:
Yep I think you are right.  Not sure what I was thinking yesterday.